macOS: Support intermediate certificates

timja opened the pull request:
Fixes https://github.com/JetBrains/jvm-native-trusted-roots/issues/3

This works now, remaining before ready for review:

* [x] Change test to not be specific to my machine
* [x] Try change to SSL policy from basic x509
* [x] Review CF objects to ensure they are being correctly released

timja committed:
macOS: Support intermediate certificates

timja committed:
macOS: Support intermediate certificates

timja commented on src/main/java/org/jetbrains/nativecerts/mac/SecurityFrameworkUtil.java:
```diff
@@ -27,7 +28,26 @@ public class SecurityFrameworkUtil {
     private final static Logger LOGGER = Logger.getLogger(SecurityFrameworkUtil.class.getName());
 
     public static List<X509Certificate> getTrustedRoots(SecurityFramework.SecTrustSettingsDomain domain) {
-        List<X509Certificate> result = SecTrustSettingsCopyCertificates(domain, cert -> isTrustedRoot(domain, cert));
```
`SecTrustSettingsCopyCertificates` is not what we want as it only copies certificates with a trust setting which isn't required

shalupov replied:
So, intermediate certificates are not returned by SecTrustSettingsCopyCertificates?

timja replied:
No as they have no trusted settings, `SecItemCopyMatching` is what is needed to find them

timja commented on src/main/java/org/jetbrains/nativecerts/mac/SecurityFrameworkUtil.java:
```diff
@@ -43,6 +63,19 @@ public static List<X509Certificate> getTrustedRoots(SecurityFramework.SecTrustSe
+    ) {
+        CFArrayRefByReference returnedCertArray = new CFArrayRefByReference();
+        SecurityFramework.OSStatus rc = SecurityFramework.INSTANCE.SecItemCopyMatching(query, returnedCertArray);
+        if (!SecurityFramework.OSStatus.errSecSuccess.equals(rc)) {
```
currently getting `One or more parameters passed to a function were not valid.`

API signature: https://developer.apple.com/documentation/security/secitemcopymatching(_:_:)?language=objc

Working native code:

```objective-c
    CFTypeRef searchKeys[] = { kSecClass, kSecMatchLimit, kSecReturnRef, kSecMatchSearchList };
    CFTypeRef searchValues[] = { kSecClassCertificate, kSecMatchLimitAll, kCFBooleanTrue, keychainList };
    search = CFDictionaryCreate(kCFAllocatorDefault,
                              searchKeys, searchValues, 4, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
    if (search == NULL) {
        goto errOut;
    }

    if( SecItemCopyMatching( search, (CFTypeRef *)&currAnchors ) == errSecSuccess ){
```

timja replied:
Hi @shalupov

I see you've done most of the work on this library.

Any chance you have a pointer to where I'm going wrong here.

Thanks!

shalupov replied:
keychainList should not be a constant here. https://github.com/openjdk/jdk/blob/59ceebc4acf2c39c86c45b1aa6746bf619417cea/src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m#L558 KeystoreImpl gets keychain list from some predefined list, but I'm worrying that it's deprecated and also does not get certificates from other domains

(some random forum message) https://developer.apple.com/forums/thread/739774

timja replied:
oh that was silly of me...

I don't think its needed anyway, I didn't implement that in node.js:

https://github.com/nodejs/node/pull/56599

timja replied:
removing it doesn't make a difference, still get the useless `One or more parameters passed to a function were not valid`

timja commented on src/test/java/org/jetbrains/nativecerts/mac/SecurityFrameworkUtilTest.java:
```diff
@@ -87,6 +93,17 @@ public void addRealUserTrustedCertificate_deny() throws Exception {
+
+        List<String> aliases = rootsAfter.stream().map(crt -> crt.getSubjectDN().toString())
+                .collect(Collectors.toList());
+        assertThat(aliases, hasItem("CN=TIMJA-ROOT, O=TIMJA, ST=ES, C=UK"));
+        assertThat(aliases.size(), is(greaterThan(1)));
+        assertThat(aliases, hasItem("CN=TIMJA-INTERMEDIATE, O=TIMJA, ST=ES, C=UK"));
```
just for testing locally, I'll look into generating on the fly once the code is working

timja commented on src/main/java/org/jetbrains/nativecerts/mac/SecurityFrameworkUtil.java:
```diff
@@ -27,7 +28,26 @@ public class SecurityFrameworkUtil {
+                createCFString("kSecClassCertificate"),
+                createCFString("kSecMatchLimitAll"),
+                createCFString("kCFBooleanTrue"),
+                createCFString("keychainList")
```
this needs to be restricted to a domain at somepoint

timja committed:
Merge branch 'trunk' into verify-certificate-when-null-trust-setting

timja committed:
Remove keychain list

shalupov committed:
unblock trusted certificates

shalupov commented:
current state: I was able to call SecItemCopyMatching without errors and it even returns something. See https://github.com/JetBrains/jvm-native-trusted-roots/commit/ef3ec665ec5afe9b086f10042af8a622a0ce1058

Basically, two things:

- kCFBooleanFalse/True are not CFString, they're CFBoolean. Those unique constants could not be easily created, so I've imported them from framework symbols
- For some reason, SecItemCopyMatching do not accepted constants created via new strings, so I've also imported them from symbols

Could you please continue on PR and I'll try to help with JNA when required?

timja committed:
Merge remote-tracking branch 'origin/sh/unblock-trusted-certificates' into verify-certificate-when-null-trust-setting

timja committed:
Some progress

timja committed:
No longer crashes

timja committed:
Working

timja commented:
Thanks for the help, its working now, I'll take a look at tests and any refactoring needed later

timja commented on .idea/misc.xml:
```diff
@@ -8,7 +8,7 @@
     </option>
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_11" default="true" project-jdk-name="11" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_17" default="true" project-jdk-name="zulu-17" project-jdk-type="JavaSDK">
```
not really sure that this should be checked in, but the project now requires Java 17:
https://github.com/JetBrains/jvm-native-trusted-roots/blob/trunk/pom.xml#L107

Tests don't run in the IDE without changing this file

shalupov replied:
Yes, it should be

shalupov replied:
In trunk: https://github.com/JetBrains/jvm-native-trusted-roots/commit/c44be3a94a9c9c6dc281f1f2333eb5bcc5117f2d

timja commented on src/main/java/org/jetbrains/nativecerts/mac/SecurityFrameworkUtil.java:
```diff
@@ -118,6 +125,34 @@ static boolean isSelfSignedCertificate(X509Certificate certificate) {
+
+            SecurityFramework.SecTrustRefByReference secTrustRefByReference = new SecurityFramework.SecTrustRefByReference();
+
+            policy = SecurityFramework.INSTANCE.SecPolicyCreateBasicX509();
```
Ideally I'll change this to an SSL policy rather than just the basic x509, this one is easier as there's a built-in function for it.

timja committed:
Add javadoc

timja committed:
Switch to SSL policy

timja committed:
Add new lines to files

timja committed:
Add some CFRelease's

timja committed:
CFRelease more

timja committed:
Generate intermediate CA

timja committed:
Implement test

shalupov commented on src/main/java/org/jetbrains/nativecerts/mac/SecurityFrameworkUtil.java:
```diff
@@ -118,31 +133,80 @@ static boolean isSelfSignedCertificate(X509Certificate certificate) {
+                return false;
+            }
+
+            return SecurityFramework.INSTANCE.SecTrustEvaluateWithError(secTrustRefByReference.getSecTrustRef(), null);
```
Could you please add a WARN message that certificated is not trusted thus not imported

timja replied:
Its valid for a certificate to be not trusted, its logged elsewhere:

```
LOGGER.fine("Certificate '" + certificateDescription + "' has no trust settings and failed to validate against trusted roots");
```

shalupov commented on src/main/java/org/jetbrains/nativecerts/mac/SecurityFrameworkUtil.java:
```diff
@@ -118,31 +133,80 @@ static boolean isSelfSignedCertificate(X509Certificate certificate) {
+            policy = SecurityFramework.INSTANCE.SecPolicyCreateSSL(false, null);
+            SecurityFramework.OSStatus rc = SecurityFramework.INSTANCE.SecTrustCreateWithCertificates(
+                    subjCerts, policy, secTrustRefByReference);
+            if (!SecurityFramework.OSStatus.errSecSuccess.equals(rc)) {
```
Could you please add a WARN message?

shalupov commented:
From failing tests, looks like system certificates are not imported

timja committed:
Merge branch 'trunk' into verify-certificate-when-null-trust-setting

timja commented:
> From failing tests, looks like system certificates are not imported

Right (annoying github action not trusting me -.-)

From stackoverflow:
https://stackoverflow.com/a/35095195

OpenJDK took the above approach:
https://github.com/openjdk/jdk/blob/master/src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m#L525-L527

Although `SecKeychainOpen` is deprecated.

This looks like the supported way to do it though:
https://developer.apple.com/documentation/security/sectrustcopyanchorcertificates(_:)?language=objc

I'll give `SecTrustCopyAnchorCertificates` a go.

timja committed:
Read from system key chain directly (not working)

timja commented:
I implemented `SecTrustCopyAnchorCertificates` but it contains non system certificates which is why openjdk went the other route.

I've implemented it using the key chain file system APIs. deprecated but still seems to be the recommended approach if you need to do this.

timja committed:
Working

timja committed:
Improve

timja committed:
Simplify

timja committed:
Add more javadoc

shalupov commented:
squashed and pushed manually https://github.com/JetBrains/jvm-native-trusted-roots/commit/477763d0bb5594f0cdff9635da9f18f52ecad674

shalupov commented:
also added more checks: https://github.com/JetBrains/jvm-native-trusted-roots/commit/96702ad6b48639c34152060b33e8bd2f0198364c
